### ISSUE TITLE
Provide a way to specify conversion fuctions for custom types for fennel.view

### ISSUE THREAD

**Issue by p-himik** [bakpakin/Fennel#356]
E.g. `awful.key` from Awesome WM Lua API creates an object of a custom type where `type(k)` is `"key"`, so using `(print (fennel.view k))` prints something like:
```
#<key: 0x563c7303a538>
```
At the same time, a key instance has properties that I know about and that I can access. It would be nice to have a way to provide a custom conversion to `fennel.view` from a custom type into one of the Lua types to allow pretty printing of instances of such types.

**Comment by technomancy** [comment]
Sure; you can do that by creating a __fennelview metamethod for
it. Presumably there's already a metatable (since it seems to be able to
override tostring) so just adding a function to that should do the trick.

Check out the docstring here and let me know if there's anything still unclear:

https://git.sr.ht/~technomancy/fennel/tree/main/src/fennel/view.fnl#L284

**Comment by p-himik** [comment]
But that's the thing - it's not only impossible to use `__fennelview` here but it's also impractical.

Why it's impossible: `bad argument #1 to 'setmetatable' (table expected, got userdata)`. As I mentioned, it's a C type exposed to Lua.
Even if it were possible to set metadata on `userdata`, it still wouldn't work because `pp-view` checks for `(type x)`, and then uses `__fennelview` iff the type is explicitly either `table` or `userdata`. But in this case `(type x)` returns `"key"`.

Why it's impractical: `fennel.view` is particularly useful to inspect the data I know little about. Setting metadata on a specific object in the hierarchy assumes that I know where that object is or requires me to walk the whole hierarchy.

Those two reasons are why I'm proposing something like [this](https://github.com/kikito/inspect.lua#optionsprocess) to be implemented in `fennel.view` as well.

**Comment by technomancy** [comment]
Oh I see; I didn't catch that you were talking about userdata specifically.

I've never used userdata but it sounds like the current provisions for
overriding are insufficient for that.

The current implementation of fennelview seems to imply that it's
possible for userdata to have metatables; do you mean that it's possible
to set metatables for userdata, but only from C, not from Lua?

In any case I think you're right that our current approach won't do. Some kind of unknown-type handler is needed.

**Comment by p-himik** [comment]
> do you mean that it's possible to set metatables for userdata, but only from C, not from Lua?

It seems like it but I'm not exactly sure since my knowledge of Lua's C API is very superficial: https://www.lua.org/pil/28.2.html

**Comment by atweiden** [comment]
As someone who has written quite a bit of Rust/Fennel using the Lua C API, extending fennelview per [kikito/inspect.lua] could only slightly improve custom userdata stringification in cases where upstream provides a reasonable definition of `tostring` for its userdata.

E.g. here’s the default output of a userdata type of mine sans `__fennelview` implementation:

```
#<Simple(
    Simple {
        env: SimpleEnv {
            protein: 1.2,
            carbs: 2.4,
            fiber: None,
            fat: 0.2,
            alcohol: None,
            per: None,
        },
    },
)>
```

It seems good enough to me. If I wanted to print other `Simple` userdata attributes here, I would modify its `tostring` implementation in the Lua C API. Or if I weren’t satisfied with the `#< ... >` wrapper fennelview adds around `Simple`’s stringification function output, I would add a `__fennelview` implementation to `Simple`’s metatable.

AwesomeWM could rather easily return something nicer than the mere stringified classname in [luaA_tolstring] (see also: [luaA_typename]).

This isn’t to say extending fennelview couldn’t greatly benefit other potential use cases — including this one, assuming patching upstream is overly cumbersome for whatever reason — just that in the specific case of *userdata* pretty printing, I haven’t yet come across anything which downstream users are in a better position to handle than upstream.

[kikito/inspect.lua]: https://github.com/kikito/inspect.lua#optionsprocess
[luaA_tolstring]: https://github.com/awesomeWM/awesome/blob/a4572b9b52d89369ce3bd462904d536ec116dc35/luaa.c#L899
[luaA_typename]: https://github.com/awesomeWM/awesome/blob/a4572b9b52d89369ce3bd462904d536ec116dc35/common/luaclass.c#L113

**Comment by p-himik** [comment]
> I haven’t yet come across anything which downstream users are in a better position to handle than upstream.

Isn't that exactly my situation? It's trivial to describe how to create a plain Lua table from an Awesome WM's `key` userdata. It not that trivial for me to patch the C implementation, rebuild Awesome, and re-run my code on the patched version.
Easy data exploration in general is possible only if you have all the means to create the representations that you want, without waiting for someone else and without having to patch someone else's code.

**Comment by atweiden** [comment]
> Easy data exploration in general is possible only if you have all the means to create the representations that you want, without waiting for someone else and without having to patch someone else's code.

You might be able to do that by redefining fennelview, e.g.:

```fennel
(local fennel (require :fennel))

(fn key-tostring [key]
  {:attr1 :val1
   :attr2 :val2})

(fn view [x]
  (if (= (type x) :key)
    (key-tostring x)
    (fennel.view x)))
```

**Comment by p-himik** [comment]
It won't help with nested data.

This issue isn't about what _I_ can do on _my_ end - I'm capable enough to solve that problem myself.
It's about making such data exploration practical and easily available to everyone using Fennel, just like in the already linked `inspect.lua`.

**Comment by p-himik** [comment]
FWIW, this is the issue that resulted in `inspect.lua`'s `process` function accepting not only the value to transform but also the path to that value: https://github.com/kikito/inspect.lua/issues/12

It doesn't seem to me that such a use-case makes an additional complexity worth it. IMO filtering keys/values and creating readable user-friendly representation are two independent functionalities.

**Comment by technomancy** [comment]
A simple implementation of viewing userdata that lacks a metamethod can
be done in only a couple lines.

Do you think this is a good solution to the problem? Is the
documentation descriptive enough and clear?

https://git.sr.ht/~technomancy/fennel/commit/238193b48163b04c942647116c0dc72acf76a5a4

Crucially the interface is the same as the metamethod; it passes in the
fennelview function itself and its state so that elements nested inside
the userdata can also be serialized.

**Comment by p-himik** [comment]
That looks promising but I'm afraid it's not enough.

The check in the commit compares `(type x)` to `:userdata`. But as I described initially, `(type x)` returns `:key` in the case of Awesome WM custom key type. I have no clue how Awesome WM does that exactly though.

Given that, seemingly, types can be completely arbitrary, the only solution I can think of is to call the custom function for all objects when the function is set in the options.

**Comment by technomancy** [comment]
Ah yeah, of course type won't work. But calling the function for everything doesn't make sense because without the idea of a table, it's impossible to have nesting, and thus all of fennelview becomes equivalent to just a plain tostring. Maybe just calling it for every element that's not a supported type instead?

**Comment by p-himik** [comment]
> without the idea of a table, it's impossible to have nesting

Not sure what you mean.

Suppose, the custom function that's called for every object looks like this (omitting all extra arguments for brevity):
```
(fn process [x]
  (if (= (type x) :key)
    {:key x.key :keysym x.keysym}
    x))
```

This way, absolutely everyone should be happy. If it's a key, Fennel formats it as a table that later gets turned into a string by `pp-table`. And if it's something else, the default processing will handle it.
And if a user wants to clearly see that it's a key and not a table of two items, then the third line can be replaced with
```
(.. "#<key: " (pp {:key x.key :keysym x.keysym}) ">")
```

> Maybe just calling it for every element that's not a supported type instead?

Sure, that would work as well, and I personally would be quite happy with it.
However, there's an arguable downside as compared to the above solution. Namely, calling `process` for all objects makes any kind of user-specified post-processing possible, like e.g. eliding tables that have more than 10 elements or eliding extremely long strings.

**Comment by technomancy** [comment]
Oh I see, you're talking about a function whose return value is still serialized; I thought you meant that the function would produce a string. That seems reasonable. Maybe "preprocess" instead of "process"?

Anyway if you want to take a shot at implementing it I think it's a good idea. If not I can get to it later.

**Comment by p-himik** [comment]
This is actually less trivial to implement than I thought, given that there's a cycle detection algorithm.
A C structure can (I think) hold a pointer to a Lua data structure that, in turn, can hold a pointer to that same C structure.

So I see three options:
- Ignore cycle detection for C types and let the user know about this limitation in the docstring
- Traverse the whole input data structure and replace all C structure instances with a memoized version of `preprocess`
- Rewrite the cycle detection algorithm to it uses the same cache as `preprocess`

Since this is a somewhat of a niche usage, I'd say the first option makes the most sense as the first implementation.
What do you think?

**Comment by technomancy** [comment]
> Ignore cycle detection for C types and let the user know about this limitation in the docstring

I agree this is the way to go; if you're doing something clever with fennelview then it's up to you to handle this problem.

**Comment by technomancy** [comment]
This ought to be closed by 6899199
